File: scripts/avm1.py
"""The AVM1 object model in a Pepper Flash process, layout *derived* at run time.

The earlier reverse work (Linux x86-64, libpepflashplayer.so 32.0.0.465) had
proved this layout. On Windows (pepflashplayer.dll win32-x64, same version
32.0.0.465) only half of it holds -- measured, not assumed:

                          Linux x86-64           Windows x86-64
    String   vtable       +0x00                  +0x00        same
             buffer       +0x08                  +0x08        same
             length       +0x30                  +0x30        same
    ScriptObject table    +0x30                  +0x30        same
    table    capacity     +0x08                  +0x08        same
             entries      +0x18                  +0x48        DIFFERENT
             stride       16 bytes               24 bytes     DIFFERENT
             entry        (value, key)           (value, _, key)

In other words: the objects look alike, but a table entry is 24 bytes and not
16, and the key is the third qword and not the second. A port that copied the
Linux offsets would have read values shifted by one field -- plausible, and
wrong.

So this module hard codes none of these offsets. It starts from one certainty:
the SWF is the same everywhere, so a string known from the SWF is interned in
the heap:

    string known from the SWF   e.g. "]=[]8", the obfuscated name of `world`
      -> its UTF-16LE buffer    heap scan
      -> the String object      the qword in front that points into the module
                                is the vtable
      -> the length offset      the qword that holds the length of the string
      -> the slots citing it    scan of the 8 atom encodings
      -> the entry stride       measured on the neighbouring keys
      -> the table base         first qword pointing into the module, before
                                the entries
      -> the value offset       a vote: the offset where all atoms are valid

Atom encoding, measured on this process:

    tag 0   signed integer        value = atom >> 3, arithmetic
    tag 1   float                 pointer to an 8 byte IEEE double
    tag 2   special               0x0a null/undefined, 0x12 false, 0x32 true
    tag 5   String                pointer to a String object
    tag 6   object                pointer to a ScriptObject
    tag 3   native object / MovieClip
"""
import struct

STRIDE_CANDS = tuple(range(0x08, 0x41, 0x08))
VAL_DELTA_CANDS = (-0x10, -0x08, 0x08, 0x10)
STR_BUF_CANDS = (0x08, 0x10, 0x18, 0x20, 0x00)
STR_LEN_CANDS = tuple(range(0x08, 0x80, 8))
SO_TBL_CANDS = tuple(range(0x08, 0x80, 8))
TBL_CAP_CANDS = (0x08, 0x10, 0x18)

MAX_TABLE_CAP = 1 << 16
MAX_STRING_LEN = 512

TAG_INT, TAG_DOUBLE, TAG_SPECIAL, TAG_NATIVE, TAG_STRING, TAG_OBJECT = 0, 1, 2, 3, 5, 6
ATOM_NULL, ATOM_FALSE, ATOM_TRUE = 0x0A, 0x12, 0x32


class Layout:
    """The offsets found for this process."""

    def __init__(self, module_lo=0):
        self.module_lo = module_lo
        self.str_vt = self.str_buf = self.str_len = None
        self.tbl_vt = self.tbl_cap = self.tbl_keys = None
        self.tbl_stride = self.tbl_value = None
        self.so_vt = self.so_tbl = None

    def complete(self):
        return None not in (self.str_vt, self.str_buf, self.str_len,
                            self.tbl_vt, self.tbl_cap, self.tbl_keys,
                            self.tbl_stride, self.tbl_value, self.so_tbl)

    def _rel(self, v):
        return "MODULE+0x%x" % (v - self.module_lo) if v else "?"

    def __str__(self):
        return (
            "  String        vtable %-18s buffer +0x%02x   length +0x%02x\n"
            "  ScriptObject  vtable %-18s table  +0x%02x\n"
            "  table         vtable %-18s capacity +0x%02x  entries +0x%02x\n"
            "  entry         %d bytes, key at +0x00, value at %+#04x"
            % (self._rel(self.str_vt), self.str_buf, self.str_len,
               self._rel(self.so_vt), self.so_tbl,
               self._rel(self.tbl_vt), self.tbl_cap, self.tbl_keys,
               self.tbl_stride, self.tbl_value)
        )


class Avm1:
    """An AVM1 view of a process, once the layout is derived."""

    def __init__(self, proc, module, heaps=None):
        self.p = proc
        self.lo, self.hi = module
        self.heaps = proc.regions() if heaps is None else heaps
        self.L = Layout(module[0])
        self._key_atoms = {}

    # -- primitives --------------------------------------------------------
    def in_module(self, v):
        return v is not None and self.lo <= v < self.hi

    def string_at(self, addr):
        """Decodes the String object at `addr`, or None if it is not one."""
        L = self.L
        if addr is None or self.p.u64(addr) != L.str_vt:
            return None
        buf, n = self.p.u64(addr + L.str_buf), self.p.u64(addr + L.str_len)
        if not buf or n is None or n > MAX_STRING_LEN:
            return None
        b = self.p.read(buf, n * 2)
        if not b or len(b) != n * 2:
            return None
        try:
            return b.decode("utf-16-le")
        except UnicodeDecodeError:
            return None

    def key_at(self, addr):
        """The name of the key stored at `addr`, or None."""
        return self.string_at((self.p.u64(addr) or 0) & ~7)

    # -- atom decoding -----------------------------------------------------
    @staticmethod
    def tag(atom):
        return None if atom is None else atom & 7

    @staticmethod
    def as_int(atom, bound=1 << 31):
        """Atom -> signed integer, or None if it is not a plausible one."""
        if atom is None or atom & 7:
            return None
        v = atom - (1 << 64) if atom >> 63 else atom
        v >>= 3                      # arithmetic shift: negatives work
        return v if -bound < v < bound else None

    def as_double(self, atom):
        if atom is None or atom & 7 != TAG_DOUBLE:
            return None
        b = self.p.read(atom & ~7, 8)
        return struct.unpack("<d", b)[0] if b and len(b) == 8 else None

    def as_number(self, atom):
        """Integer or float, depending on the tag."""
        d = self.as_double(atom)
        return d if d is not None else self.as_int(atom)

    @staticmethod
    def as_bool(atom):
        if atom == ATOM_TRUE:
            return True
        if atom == ATOM_FALSE:
            return False
        return None

    def as_string(self, atom):
        if atom is None or atom & 7 not in (TAG_STRING, TAG_INT):
            return None
        return self.string_at(atom & ~7)

    def well_formed(self, atom):
        """Can the atom be read? Used by the vote on the value offset."""
        if atom is None:
            return False
        tag = atom & 7
        if tag == TAG_SPECIAL:
            return atom in (ATOM_NULL, ATOM_FALSE, ATOM_TRUE)
        if tag == TAG_DOUBLE:
            return self.as_double(atom) is not None
        if tag == TAG_STRING:
            return self.string_at(atom & ~7) is not None
        if tag in (TAG_OBJECT, TAG_NATIVE):
            return self.in_module(self.p.u64(atom & ~7))
        if tag == TAG_INT:
            return self.as_int(atom) is not None
        return False

    def describe(self, atom):
        if atom is None:
            return "<unreadable>"
        tag = atom & 7
        if tag == TAG_SPECIAL:
            return {ATOM_NULL: "null", ATOM_FALSE: "false",
                    ATOM_TRUE: "true"}.get(atom, "special 0x%x" % atom)
        if tag == TAG_DOUBLE:
            return "float %r" % self.as_double(atom)
        if tag == TAG_STRING:
            return "String %r" % self.string_at(atom & ~7)
        if tag in (TAG_OBJECT, TAG_NATIVE):
            return "objet 0x%x" % (atom & ~7)
        v = self.as_int(atom)
        return "int %d" % v if v is not None else "0x%x" % atom

    # -- tables ------------------------------------------------------------
    def capacity(self, tbl):
        cap = self.p.u64(tbl + self.L.tbl_cap)
        return cap if cap and 0 < cap <= MAX_TABLE_CAP else None

    def entries(self, tbl):
        """[(key name, value atom, key address)] for the table."""
        cap = self.capacity(tbl)
        if cap is None:
            return []
        L, out = self.L, []
        for i in range(cap):
            k = tbl + L.tbl_keys + i * L.tbl_stride
            name = self.key_at(k)
            if name is not None:
                out.append((name, self.p.u64(k + L.tbl_value), k))
        return out

    def slot(self, tbl, key):
        """The address of the *value* slot of `key`, or None.

        The keys are interned: once the atom is known, the next search is an
        integer comparison, not a string decode.
        """
        cap = self.capacity(tbl)
        if cap is None:
            return None
        L = self.L
        want = self._key_atoms.get(key)
        for i in range(cap):
            k = tbl + L.tbl_keys + i * L.tbl_stride
            raw = self.p.u64(k)
            if not raw:
                continue
            if want is not None:
                if raw == want:
                    return k + L.tbl_value
            elif self.string_at(raw & ~7) == key:
                self._key_atoms[key] = raw
                return k + L.tbl_value
        return None

    def get(self, tbl, key):
        s = self.slot(tbl, key)
        return None if s is None else self.p.u64(s)

    def table_of(self, atom):
        """Object atom -> its property table, or None."""
        if (atom is None or self.L.so_tbl is None
                or atom & 7 not in (TAG_OBJECT, TAG_NATIVE, TAG_INT)):
            return None
        so = atom & ~7
        if not self.in_module(self.p.u64(so)):
            return None
        t = self.p.u64(so + self.L.so_tbl)
        return t if t and self.p.u64(t) == self.L.tbl_vt else None

    def child(self, tbl, key):
        """Shortcut: the table of the object stored under `key`."""
        return self.table_of(self.get(tbl, key))

    # -- layout derivation -------------------------------------------------
    def derive(self, anchor, verbose=False):
        """Finds every offset from one key known from the SWF.

        -> the list of tables that own this key.
        """
        log = print if verbose else (lambda *_: None)
        strobj = self._derive_string(anchor)
        if strobj is None:
            log("  string %r not found in the heap" % anchor)
            return []
        log("  String %r at 0x%x: vtable %s, buffer +0x%02x, length +0x%02x"
            % (anchor, strobj, self.L._rel(self.L.str_vt),
               self.L.str_buf, self.L.str_len))

        tables = []
        for ks in self.p.scan_tagged(strobj, regions=self.heaps):
            if self.key_at(ks) != anchor:
                continue
            t = self._derive_table(ks)
            if t is not None and t not in tables:
                tables.append(t)
                log("  table 0x%x: capacity %d, %d readable entries"
                    % (t, self.capacity(t) or 0, len(self.entries(t))))
        return tables

    def _derive_string(self, text):
        """Finds the String object of `text` and derives the String layout."""
        pat = text.encode("utf-16-le")
        for buf in self.p.scan(pat, align=2, regions=self.heaps):
            for ref in self.p.scan(struct.pack("<Q", buf), align=8,
                                   regions=self.heaps):
                for ob in STR_BUF_CANDS:
                    so, vt = ref - ob, self.p.u64(ref - ob)
                    if not self.in_module(vt):
                        continue
                    for ol in STR_LEN_CANDS:
                        if self.p.u64(so + ol) != len(text):
                            continue
                        self.L.str_vt, self.L.str_buf, self.L.str_len = vt, ob, ol
                        if self.string_at(so) == text:
                            return so
        return None

    def _derive_table(self, keyslot):
        """Derives stride, base, capacity and value offset from one key.

        The geometry is derived only once. The tables that follow are simply
        located with it. Otherwise the last table examined would force its own
        measurement on all the others.
        """
        if self.L.tbl_stride is not None:
            return self._table_base(keyslot)
        stride = self._measure_stride(keyslot)
        if stride is None:
            return None
        first = keyslot
        while self.key_at(first - stride) is not None:
            first -= stride
        n = 0
        while self.key_at(first + n * stride) is not None:
            n += 1

        for back in range(8, 0x101, 8):
            t = first - back
            if not self.in_module(self.p.u64(t)):
                continue
            for co in TBL_CAP_CANDS:
                cap = self.p.u64(t + co)
                if cap is None or not (n <= cap <= MAX_TABLE_CAP):
                    continue
                self.L.tbl_vt, self.L.tbl_cap = self.p.u64(t), co
                self.L.tbl_keys, self.L.tbl_stride = back, stride
                if self._derive_value_offset(t):
                    return t
        return None

    def _table_base(self, keyslot):
        """The base of the table holding `keyslot`, geometry already known."""
        L = self.L
        first = keyslot
        while self.key_at(first - L.tbl_stride) is not None:
            first -= L.tbl_stride
        t = first - L.tbl_keys
        return t if self.p.u64(t) == L.tbl_vt and self.capacity(t) else None

    def _measure_stride(self, keyslot):
        """The stride is the smallest gap where both neighbours are keys too."""
        for d in STRIDE_CANDS:
            if (self.key_at(keyslot + d) is not None
                    and self.key_at(keyslot + 2 * d) is not None):
                return d
        return None

    def _derive_value_offset(self, tbl):
        """A vote: the right offset is where atoms are valid AND varied.

        Validity alone is not enough. An entry holds an unused qword that is
        always zero, and zero is a perfectly valid integer atom. So that
        column scores perfectly while holding nothing. The diversity test is
        what rejects it, and nothing else.
        """
        best, best_score = None, 0
        for d in VAL_DELTA_CANDS:
            self.L.tbl_value = d
            ents = self.entries(tbl)
            if not ents:
                continue
            vals = [v for _, v, _ in ents]
            diversity = len(set(vals)) / len(vals)
            if diversity < 0.25:
                continue
            score = sum(1 for v in vals if self.well_formed(v)) / len(vals)
            # The column of the next entry holds the same values, shifted by
            # one: it differs by one atom at most, and can win by that atom
            # alone. So a later candidate must win by more than one entry.
            if score > best_score + 1 / len(vals):
                best, best_score = d, score
        self.L.tbl_value = best
        return best is not None and best_score > 0.95

    def derive_script_object(self, atom, expect_key=None):
        """Derives so_vt / so_tbl from an atom known to be an object.

        `expect_key` is a property the target object certainly owns. Without
        it, several offsets can lead to a plausible table, and the wrong one
        would then be cached for every read that follows. So we demand the
        constraint whenever we have one.
        """
        so = (atom or 0) & ~7
        vt = self.p.u64(so)
        if not self.in_module(vt):
            return None
        for x in SO_TBL_CANDS:
            t = self.p.u64(so + x)
            if not t or self.p.u64(t) != self.L.tbl_vt or not self.capacity(t):
                continue
            if expect_key is not None and self.slot(t, expect_key) is None:
                continue
            self.L.so_vt, self.L.so_tbl = vt, x
            return t
        return None

File: src/reader/avm1.rs
//! The AVM1 object model inside a Pepper Flash process.
//!
//! See `docs/reverse-engineering.md` for the measurements. In short: String
//! and ScriptObject have the same layout on Linux and on Windows, but
//! property tables do not.
//!
//! ```text
//!                   Linux x86-64        Windows x86-64
//!   entries         tbl+0x18            tbl+0x48
//!   stride          16 bytes            24 bytes
//!   entry           (value, key)        (value, _, key)
//! ```
//!
//! Both forms are tried, and a semantic check decides between them: a profile
//! that does not lead to a known Hammerfest world is rejected. On failure we
//! return nothing. We never return a wrong level.
//!
//! The vtables are not hard coded. They are found at run time, from a string
//! we know the SWF contains.

// Atom decoding lives in the core, where it is tested.
use hammerfest_core::atom;
pub use hammerfest_core::atom::{as_bool, as_int};

/// Everything the reader needs from a process.
///
/// One method, because the reader makes exactly four kinds of raw read and
/// all four are "give me these bytes". The contract is ours, so a test double
/// implements it without borrowing a word of the runtime ABI.
///
/// `read_into` fills `buf` whole, or returns `None`. It never fills a part of
/// it. The caller then has no partially true buffer to mistake for a value,
/// which is what `reader::refuses-rather-than-defaults` asks for.
///
/// `tests/contract/memory.rs` states that in four questions. The test heap and
/// the production adapter, `ProcessMemory`, both answer them.
///
/// @spec reader::refuses-rather-than-defaults
pub trait Memory {
    fn read_into(&self, address: u64, buf: &mut [u8]) -> Option<()>;
}

/// Atom -> number, integer or float.
///
/// `duration` is the integer 0 when the GameMode is built, then becomes a
/// float on the first frame played. Both forms are normal. Reading only one of
/// them would mean reading nothing during the black screen.
pub fn as_number(mem: &dyn Memory, atom: u64) -> Option<f64> {
    match atom::double_at(atom) {
        Some(addr) => read_u64(mem, addr).map(atom::decode_double),
        None => as_int(atom).map(|v| v as f64),
    }
}

/// The width of a pointer, of a length and of an atom in a build of Flash
/// Player.
///
/// The object model is the same on both widths. Measured on the Windows
/// projector 32.0.0.465, every offset of a 32-bit build is the one of Pepper
/// Flash under Linux counted in words of four bytes instead of eight.
#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
pub enum Word {
    /// A 32-bit build: the Windows projector.
    Four,
    /// A 64-bit build: Pepper Flash, and the projector under Linux.
    #[default]
    Eight,
}

impl Word {
    pub const fn bytes(self) -> u64 {
        match self {
            Word::Four => 4,
            Word::Eight => 8,
        }
    }

    /// The word at `addr`: a pointer, a length or a capacity.
    #[inline]
    pub fn read(self, mem: &dyn Memory, addr: u64) -> Option<u64> {
        match self {
            Word::Eight => read_u64(mem, addr),
            Word::Four => read_u32(mem, addr),
        }
    }

    /// The atom at `addr`, in the 64-bit form the core decodes.
    ///
    /// An integer keeps its sign: `portalId` is -1, `0xfffffff8` in four
    /// bytes.
    ///
    /// @spec projector::words-of-four-bytes
    #[inline]
    pub fn atom(self, mem: &dyn Memory, addr: u64) -> Option<u64> {
        let raw = self.read(mem, addr)?;
        let is_short_integer = self == Word::Four && atom::tag(raw) == atom::TAG_INT;
        Some(if is_short_integer {
            raw as u32 as i32 as i64 as u64
        } else {
            raw
        })
    }

    /// The word at `offset` in a buffer already read, if it fits there.
    pub fn in_buffer(self, buf: &[u8], offset: usize) -> Option<u64> {
        crate::scan::word_at(buf, offset, self.bytes() as usize)
    }

    /// The table geometries a build of this width may use.
    pub fn profiles(self) -> &'static [Profile] {
        match self {
            Word::Four => PROFILES_32_BITS,
            Word::Eight => PROFILES,
        }
    }

    /// Candidate offsets for the buffer pointer of a String object.
    pub fn string_buffer_candidates(self) -> [u64; 5] {
        let word = self.bytes();
        [word, 2 * word, 3 * word, 4 * word, 0]
    }

    /// Candidate offsets for `ScriptObject -> table`, from one word to
    /// fifteen. Six words first: it is the offset on every known build, but
    /// we still derive it.
    fn table_pointer_candidates(self) -> [u64; 15] {
        const MEASURED_WORDS: u64 = 6;
        let word = self.bytes();
        let mut candidates = [MEASURED_WORDS * word; 15];
        let others = (1..=15).filter(|&words| words != MEASURED_WORDS);
        for (candidate, words) in candidates[1..].iter_mut().zip(others) {
            *candidate = words * word;
        }
        candidates
    }
}
/// Entries searched back from a key, looking for the header of its table.
///
/// The header is `i * stride + keys` bytes in front of entry `i`, and `i` is
/// unknown. Measured on a running game, the anchor keys sat at entries 2, 61,
/// 70, 97 and 570 of their tables, the largest of which held 793 entries.
///
/// The rule used to be "walk back while the previous qword decodes as a
/// string, then subtract the key offset". It held on Windows and it does not
/// hold on macOS: the heap there packs String pointers on both sides of a
/// table, so the walk ran up to 95 entries past the base and never landed on
/// a header.
///
/// It does not hold on Linux either, for the opposite reason: a table there
/// can hold a key that is not a String object. Measured on a running game, the
/// entry in front of `world` held one, so the walk stopped short of the base.
const MAX_BACK: u64 = 2048;
/// Bytes read at a time while searching back.
const BACK_CHUNK: usize = 8192;
const MAX_CAPACITY: u64 = 1 << 16;
/// Maximum length of a decoded key. Obfuscated names are 2 to 8 characters.
const MAX_KEY: usize = 64;

/// The geometry of a property table.
#[derive(Copy, Clone, Debug)]
pub struct Profile {
    pub name: &'static str,
    /// Offset of the first key from the base of the table.
    pub keys: u64,
    /// Size of one entry.
    pub stride: u64,
    /// Position of the value, relative to the key.
    pub value: i64,
}

pub const PROFILES: &[Profile] = &[
    Profile {
        name: "windows-x64",
        keys: 0x58,
        stride: 24,
        value: -0x10,
    },
    Profile {
        name: "linux-x64",
        keys: 0x20,
        stride: 16,
        value: -0x08,
    },
];

/// The geometry of a property table on the Windows projector, measured under
/// Wine: the one of `linux-x64`, in words of four bytes.
pub const PROFILES_32_BITS: &[Profile] = &[Profile {
    name: "windows-x86",
    keys: 0x10,
    stride: 8,
    value: -0x04,
}];

#[derive(Copy, Clone, Debug)]
pub struct Layout {
    pub module: (u64, u64),
    pub str_vt: u64,
    pub str_buf: u64,
    pub str_len: u64,
    pub tbl_vt: u64,
    pub profile: Profile,
    pub so_tbl: u64,
    pub word: Word,
}

/// Both the Flash player and the host are x86-64, so the bytes are little
/// endian on each side of this call.
#[inline]
pub fn read_u64(mem: &dyn Memory, addr: u64) -> Option<u64> {
    if addr == 0 || addr >= 1 << 47 {
        return None;
    }
    let mut bytes = [0u8; 8];
    mem.read_into(addr, &mut bytes)?;
    Some(u64::from_le_bytes(bytes))
}

/// A 32-bit word, for a 32-bit build.
#[inline]
fn read_u32(mem: &dyn Memory, addr: u64) -> Option<u64> {
    if addr == 0 || addr >= 1 << 47 {
        return None;
    }
    let mut bytes = [0u8; 4];
    mem.read_into(addr, &mut bytes)?;
    Some(u64::from(u32::from_le_bytes(bytes)))
}

impl Layout {
    #[inline]
    pub fn in_module(&self, v: u64) -> bool {
        v >= self.module.0 && v < self.module.1
    }

    /// Reads a String object into `out`, and returns the number of UTF-16 units.
    pub fn read_string(
        &self,
        mem: &dyn Memory,
        addr: u64,
        out: &mut [u16; MAX_KEY],
    ) -> Option<usize> {
        if self.word.read(mem, addr)? != self.str_vt {
            return None;
        }
        let buf = self.word.read(mem, addr + self.str_buf)?;
        let n = self.word.read(mem, addr + self.str_len)? as usize;
        if n == 0 || n > MAX_KEY {
            return None;
        }
        let mut bytes = [0u8; MAX_KEY * 2];
        let bytes = &mut bytes[..n * 2];
        mem.read_into(buf, bytes)?;
        let (pairs, _) = bytes.as_chunks::<2>();
        for (unit, pair) in out[..n].iter_mut().zip(pairs) {
            *unit = u16::from_le_bytes(*pair);
        }
        Some(n)
    }

    /// Is the string at `addr` exactly `want`?
    ///
    /// The comparison allocates nothing. The keys are known at compile time,
    /// so we encode `want` to UTF-16 as we go.
    pub fn string_eq(&self, mem: &dyn Memory, addr: u64, want: &str) -> bool {
        let mut buf = [0u16; MAX_KEY];
        let Some(n) = self.read_string(mem, addr, &mut buf) else {
            return false;
        };
        let mut it = want.encode_utf16();
        for &c in &buf[..n] {
            if it.next() != Some(c) {
                return false;
            }
        }
        it.next().is_none()
    }

    // -- tables ------------------------------------------------------------
    pub fn capacity(&self, mem: &dyn Memory, tbl: u64) -> Option<u64> {
        let cap = self.word.read(mem, tbl + self.capacity_offset())?;
        (cap > 0 && cap <= MAX_CAPACITY).then_some(cap)
    }

    /// Where a table keeps its capacity: right after its vtable.
    #[inline]
    fn capacity_offset(&self) -> u64 {
        self.word.bytes()
    }

    /// Address of key number `i` in the table.
    #[inline]
    pub fn key_addr(&self, tbl: u64, i: u64) -> u64 {
        tbl + self.profile.keys + i * self.profile.stride
    }

    /// The name of the key stored at `addr`. It masks the tag, because a key
    /// can be stored as a raw pointer or as an atom, depending on the
    /// platform.
    pub fn key_is(&self, mem: &dyn Memory, addr: u64, want: &str) -> bool {
        match self.word.read(mem, addr) {
            Some(raw) => self.string_eq(mem, raw & !7, want),
            None => false,
        }
    }

    /// The atom of property `key`, or None.
    pub fn get(&self, mem: &dyn Memory, tbl: u64, key: &str) -> Option<u64> {
        let mut ignored = 0;
        self.get_cached(mem, tbl, key, &mut ignored)
    }

    /// Like `get`, but it remembers the index of the entry.
    ///
    /// Walking the 128 entries of a table on every read would cost hundreds of
    /// memory accesses per tick. So we keep the index -- but we check it again
    /// before use: if the key is no longer there, we search again. The thing
    /// we must never keep is the *final* address, not the path.
    pub fn get_cached(&self, mem: &dyn Memory, tbl: u64, key: &str, hint: &mut u64) -> Option<u64> {
        let cap = self.capacity(mem, tbl)?;
        let value_at = |k: u64| self.word.atom(mem, (k as i64 + self.profile.value) as u64);

        if *hint < cap {
            let k = self.key_addr(tbl, *hint);
            if self.key_is(mem, k, key) {
                return value_at(k);
            }
        }
        for i in 0..cap {
            let k = self.key_addr(tbl, i);
            if self.key_is(mem, k, key) {
                *hint = i;
                return value_at(k);
            }
        }
        None
    }

    pub fn get_int(&self, mem: &dyn Memory, tbl: u64, key: &str) -> Option<i64> {
        as_int(self.get(mem, tbl, key)?)
    }

    /// The property table of the object that `atom` points to.
    pub fn table_of(&self, mem: &dyn Memory, atom: u64) -> Option<u64> {
        let so = atom & !7;
        if !self.in_module(self.word.read(mem, so)?) {
            return None;
        }
        let t = self.word.read(mem, so + self.so_tbl)?;
        (self.word.read(mem, t)? == self.tbl_vt).then_some(t)
    }

    /// The table of the object stored under `key`.
    pub fn child(&self, mem: &dyn Memory, tbl: u64, key: &str) -> Option<u64> {
        self.table_of(mem, self.get(mem, tbl, key)?)
    }

    pub fn child_cached(
        &self,
        mem: &dyn Memory,
        tbl: u64,
        key: &str,
        hint: &mut u64,
    ) -> Option<u64> {
        self.table_of(mem, self.get_cached(mem, tbl, key, hint)?)
    }

    /// Finds `so_tbl` from an object whose property we know.
    ///
    /// Without the `expect_key` constraint, several offsets lead to a
    /// plausible table, and the wrong one would be cached for every read that
    /// follows.
    pub fn derive_so_tbl(&mut self, mem: &dyn Memory, atom: u64, expect_key: &str) -> Option<u64> {
        let so = atom & !7;
        if !self.in_module(self.word.read(mem, so)?) {
            return None;
        }
        for off in self.word.table_pointer_candidates() {
            let Some(t) = self.word.read(mem, so + off) else {
                continue;
            };
            if self.word.read(mem, t) != Some(self.tbl_vt) {
                continue;
            }
            if self.capacity(mem, t).is_none() {
                continue;
            }
            self.so_tbl = off;
            if self.get(mem, t, expect_key).is_some() {
                return Some(t);
            }
        }
        None
    }

    /// The base of the table that contains `keyslot`, for the current profile.
    ///
    /// We walk back while the previous qword decodes as a string, then we
    /// subtract the key offset. A wrong profile gives a base whose vtable is
    /// not a vtable, so it rejects itself.
    ///
    /// If `tbl_vt` is unknown (zero), any pointer into the module will do and
    /// becomes the reference vtable. That is how it is derived instead of hard
    /// coded.
    pub fn table_base(&mut self, mem: &dyn Memory, keyslot: u64) -> Option<u64> {
        let stride = self.profile.stride;
        let mut buf = alloc::vec![0u8; BACK_CHUNK + 8];
        let mut i = 0u64;

        while i < MAX_BACK {
            // The headers of the candidates in this chunk, from the nearest
            // one down. One read covers them all.
            let n = (BACK_CHUNK as u64 / stride).min(MAX_BACK - i);
            let hi = keyslot.checked_sub(i * stride + self.profile.keys)?;
            let lo = hi.checked_sub((n - 1) * stride)?;
            let span = (hi - lo) as usize + 16;
            let chunk = mem.read_into(lo, &mut buf[..span]).map(|_| &buf[..span]);

            for k in 0..n {
                let tbl = hi - k * stride;
                // A chunk that would not read is served one address at a
                // time. A range near the edge of a region refuses as a whole,
                // and the header we want can sit inside it.
                let vt = match chunk {
                    Some(bytes) => self.word.in_buffer(bytes, (tbl - lo) as usize)?,
                    None => match self.word.read(mem, tbl) {
                        Some(vt) => vt,
                        None => continue,
                    },
                };
                if self.tbl_vt == 0 {
                    if !self.in_module(vt) {
                        continue;
                    }
                } else if vt != self.tbl_vt {
                    continue;
                }
                // A table holds our key, so its capacity must cover its index.
                let cap = match chunk {
                    Some(bytes) => self
                        .word
                        .in_buffer(bytes, (tbl - lo + self.capacity_offset()) as usize)
                        .filter(|&c| c > 0 && c <= MAX_CAPACITY),
                    None => self.capacity(mem, tbl),
                };
                let Some(cap) = cap else { continue };
                if i + k >= cap {
                    continue;
                }
                self.tbl_vt = vt;
                return Some(tbl);
            }
            i += n;
        }
        None
    }
}
